src/libfsm/trim.c

```diff
@@ -14,198 +14,332 @@
-			     state_set_next(&state_iter, &epsilon_s); ) {
-				if (marks[s] & MARK_REACHES_END) {
-					break;
-				}
-				if (marks[epsilon_s] & MARK_REACHES_END) {
-					marks[s] |= MARK_REACHES_END;
-				}
+		}
+	}
+
+	/* Only tracking reachability from start, so we're done. */
+	if (mode == FSM_TRIM_START_REACHABLE) {
+		queue_free(q);
+		return 1;
+	}
+
+	/* Clear mark for second pass. */
+	for (s_id = 0; s_id < state_count; s_id++) {
+		fsm->states[s_id].visited = 0;
+	}
+
+	/* Sort edges by state they lead to, inverting the index. */
+	qsort(edges, edge_count, sizeof(edges[0]), cmp_edges_by_to);
+
+	/* Reuse the existing queue for a second breadth-first walk.
+	 * This assumes the queue can be reused once empty; if the
+	 * implementation changes, it may need to reset the queue or
+	 * construct a new one.
+	 *
+	 * Initialize the queue with the end states. */
+	{
+		size_t e_i;
+		for (e_i = 0; e_i < end_count; e_i++) {
+			const fsm_state_t end_id = ends[e_i];
+			assert(end_id < state_count);
+			if (!queue_push(q, end_id)) {
+				goto cleanup;
 			}
+			fsm->states[end_id].visited = 1;
+		}
 
-			for (edge_set_reset(fsm->states[s].edges, &edge_iter);
-			     edge_set_next(&edge_iter, &e); ) {
-				if (marks[s] & MARK_REACHES_END) {
-					break;
-				}
-				if (marks[e.state] & MARK_REACHES_END) {
-					marks[s] |= MARK_REACHES_END;
-				}
+		/* The ends are no longer needed. */
+		f_free(fsm->opt->alloc, ends);
+		ends = NULL;
+	}
+
+	if (edge_count == 0) {	/* done */
+		res = 1;
+		goto cleanup;
+	}
+
+	/* Build the offset table into the edge index, so we
+	 * can map from s_id to the consecutive edges leading
+	 * to it with random access.
+	 *
+	 * offsets[s_id] contains the offset of the first edge after
+	 * s_id, so that offsets[s_id - 1] can be used as the starting
+	 * offset, or 0 when s_id is 0. Since states may not appear in
+	 * the table, any case where offsets[i] == 0 is set to
+	 * offsets[i - 1], to represent zero entries. */
+	{
+		size_t i;
+		const fsm_state_t max_to = edges[edge_count - 1].to;
+		offsets = f_calloc(fsm->opt->alloc,
+		    (max_to + 1), sizeof(offsets[0]));
+		if (offsets == NULL) {
+			goto cleanup;
+		}
+
+		for (i = 0; i < edge_count; i++) {
+			const fsm_state_t to = edges[i].to;
+			offsets[to] = i + 1;
+		}
+
+		for (i = 0; i <= max_to; i++) { /* fill in gaps */
+			if (i > 0 && offsets[i] == 0) {
+				offsets[i] = offsets[i - 1];
 			}
+		}
+		assert(offsets[max_to] == edge_count);
+	}
+
 
-			marks[s] |= MARK_DONE;
-			stack_i--;
-		} else {
-			assert(!"unreachable");
+	/* Walk breadth-first from ends and mark reachable states. */
+	while (queue_pop(q, &s_id)) {
+		size_t base, limit, e_i;
+		assert(fsm->states[s_id].visited);
+
+		base = (s_id == 0 ? 0 : offsets[s_id - 1]);
+		limit = offsets[s_id];
+		for (e_i = base; e_i < limit; e_i++) {
+			const fsm_state_t from = edges[e_i].from;
+			assert(from < state_count);
+			if (!fsm->states[from].visited) {
+				fsm->states[from].visited = 1;
```

Comment:
It might be worth (optionally) save the min distance from any end state.

I believe states that are different distances to their closest end state cannot be equivalent, so you could seed Hopcroft with partitions based on the distance to closest end state.